src/fu-engine.c

```diff
@@ -1815,6 +1821,86 @@ fu_engine_appstream_upgrade_cb (XbBuilderFixup *self,
 	return TRUE;
 }
 
+static XbBuilderSource *
+fu_engine_create_metadata_builder_source (FuEngine *self,
+					  const gchar *fn,
+					  GError **error)
+{
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(XbSilo) silo = NULL;
+	g_autoptr(XbBuilderSource) source = xb_builder_source_new ();
+	g_autofree gchar *xml = NULL;
+
+	g_debug ("building metadata for %s", fn);
+	blob = fu_common_get_contents_bytes (fn, error);
+	if (blob == NULL)
+		return NULL;
+
+	/* convert the silo for the CAB into a XbBuilderSource */
+	silo = fu_engine_get_silo_from_blob (self, blob, error);
+	if (silo == NULL)
+		return NULL;
+	xml = xb_silo_export (silo, XB_NODE_EXPORT_FLAG_NONE, error);
+	if (xml == NULL)
+		return NULL;
+	if (!xb_builder_source_load_xml (source, xml,
+					 XB_BUILDER_SOURCE_FLAG_NONE,
+					 error))
+		return NULL;
+	return g_steal_pointer (&source);
+}
+
+static gboolean
+fu_engine_create_metadata (FuEngine *self, XbBuilder *builder,
+			   FwupdRemote *remote, GError **error)
+{
+	g_autoptr(GPtrArray) files = NULL;
+	g_autofree gchar *pattern = NULL;
+	const gchar *path;
+
+	/* find all files in directory */
+	path = fwupd_remote_get_filename_cache (remote);
+	files = fu_common_get_files_recursive (path, error);
+	if (files == NULL)
+		return FALSE;
+
+	/* add each source */
+	pattern = g_build_filename (path, "*.cab", NULL);
+	for (guint i = 0; i < files->len; i++) {
+		g_autoptr(XbBuilderNode) custom = NULL;
+		g_autoptr(XbBuilderSource) source = NULL;
+		g_autoptr(GError) error_local = NULL;
+		const gchar *fn = g_ptr_array_index (files, i);
+
+		/* check is cab file */
+		if (fnmatch (pattern, fn, 0) != 0) {
```

Comment:
Was there a reason to use `fnmatch` rather than just `g_str_has_suffix(fn, ".cab")`?